File: Scripts/Script_stations.py
import json


def preprocess_json(input_json_file, output_json_file):
    try:
        with open(input_json_file, 'r', encoding='utf-8') as file:
            original_data = json.load(file)
    except Exception as e:
        print(f"Error reading or loading the JSON file: {e}")
        exit(1)

    transformed_data = []
    for English_name, details in original_data.items():
        new_entry = {
            "English_name": English_name
        }
        new_entry.update(details)
        transformed_data.append(new_entry)

    try:
        with open(output_json_file, 'w', encoding='utf-8') as file:
            json.dump(transformed_data, file, ensure_ascii=False, indent=4)
    except Exception as e:
        print(f"Error writing output file: {e}")
        exit(1)


def generate_sql(input_json_file, output_sql_file):
    create_table = """
        create table if not exists stations
        (
            station_id serial  primary key,
            Chinese_name varchar(50) not null unique,
            English_name varchar(50) not null unique,
            district varchar(50) not null,
            intro varchar(5000) not null
        );
        """

    try:
        with open(input_json_file, 'r', encoding='utf-8') as file:
            data = json.load(file)
    except Exception as e:
        print(f"Error reading or loading the JSON file: {e}")
        exit(1)

    try:
        with open(output_sql_file, 'w', encoding='utf-8') as file:
            file.write(create_table + "\n")
            for record in data:
                english_name = record['English_name'].replace("'", "''").replace("\n", "")
                chinese_name = record['chinese_name'].replace("'", "''")
                intro = record['intro'].replace("'", "''")
                district = record['district']

                insert_sql = (f"INSERT INTO stations (Chinese_name, English_name, district, intro) "
                              f"VALUES ('{chinese_name}', '{english_name}', '{district}', '{intro}');\n")
                file.write(insert_sql)
    except Exception as e:
        print(f"Error with reading the JSON file: {e}")


json_input_path = '../resource/stations.json'  # original JSON file path
json_output_path = '../out/stations_adjusted.json'  # adjusted JSON file path
sql_output_path = '../out/stations.sql'  # SQL file path

preprocess_json(json_input_path, json_output_path)
generate_sql(json_output_path, sql_output_path)

File: Scripts_in_Java/Script_stations.java
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonWriter;

import java.io.*;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class Script_stations {

    public static void preprocessJson(String inputJsonFile, String outputJsonFile) {
        try (Reader reader = new FileReader(inputJsonFile)) {
            Type type = new TypeToken<Map<String, Map<String, Object>>>(){}.getType();
            Map<String, Map<String, Object>> originalData = new Gson().fromJson(reader, type);
            List<Map<String, Object>> transformedData = new ArrayList<>();

            originalData.forEach((englishName, details) -> {
                details.put("English_name", englishName);
                transformedData.add(details);
            });

            try (Writer writer = new FileWriter(outputJsonFile);
                 JsonWriter jsonWriter = new JsonWriter(writer)) {
                jsonWriter.setIndent("    ");
                new Gson().toJson(transformedData, List.class, jsonWriter);
            }

        } catch (IOException e) {
            System.out.println("Error reading or writing JSON file: " + e.getMessage());
        }
    }

    public static void generateSql(String inputJsonFile, String outputSqlFile) {
        String createTable = """
            create table if not exists stations
            (
                station_id serial primary key,
                Chinese_name varchar(50) not null unique,
                English_name varchar(50) not null unique,
                district varchar(50) not null,
                intro varchar(5000) not null
            );
            """;

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputSqlFile));
             Reader reader = new FileReader(inputJsonFile)) {
            writer.write(createTable + "\n");
            Type listType = new TypeToken<List<Map<String, Object>>>(){}.getType();
            List<Map<String, Object>> data = new Gson().fromJson(reader, listType);

            data.forEach(record -> {
                try {
                    String English_name = ((String) record.get("English_name")).replace("'", "''").replace("\n", "");
                    String Chinese_name = ((String) record.get("chinese_name")).replace("'", "''");
                    String intro = ((String) record.get("intro")).replace("'", "''");
                    String district = (String) record.get("district");

                    String insertSql = String.format("INSERT INTO stations (Chinese_name, English_name, district, intro) VALUES ('%s', '%s', '%s', '%s');\n",
                            Chinese_name, English_name, district, intro);
                    writer.write(insertSql);
                } catch (IOException e) {
                    System.out.println("Error reading or writing JSON file: " + e.getMessage());
                }
            });
        } catch (IOException e) {
            System.out.println("Error reading or writing JSON file: " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        String jsonInputPath = "C:\\Users\\Eric\\IdeaProjects\\CS307\\Project1\\resource\\stations.json";
        String jsonOutputPath = "C:\\Users\\Eric\\IdeaProjects\\CS307\\Project1\\out\\stations_adjusted.json";
        String sqlOutputPath = "C:\\Users\\Eric\\IdeaProjects\\CS307\\Project1\\out\\stations.sql";

        preprocessJson(jsonInputPath, jsonOutputPath);
        generateSql(jsonOutputPath, sqlOutputPath);
    }
}
